### PR TITLE
feat: nonce validation config

### PR DESCRIPTION
**Pull request by greged93**
Add nonce validation configuration.
# Pull Request type

Please add the labels corresponding to the type of changes your PR introduces:

- Feature
- Testing

## What is the current behavior?
Nonce validation is always applied.

Resolves: #846 

## What is the new behavior?
Nonce validation is set by default but can be deactivated.

- Add nonce validation enabling in the config.
- Test nonce validation config.

## Does this introduce a breaking change?
Yes

### CODE REVIEW

**Comment by greged93** [comment]
Nonce ordering is still kept in the `validate_unsigned` call. Can be skipped if nonce validation is disabled but for me seems normal to keep it.

File: crates/pallets/starknet/src/tests/constants.rs [hunk 1]
```diff
@@ -31,3 +31,5 @@ pub const EMIT_SINGLE_EVENT_CLASS_HASH: &str = "0x91000";
 pub const MULTIPLE_EVENT_EMITTING_CONTRACT_ADDRESS: &str =
     "0x051a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02cf";
 pub const MULTIPLE_EVENT_EMITTING_CLASS_HASH: &str = "0x92000";
+
+pub const SALT: &str = "0x03b37cbe4e9eac89d54c5f7cc6329a63a63e8c8db2bf936f981041e086752463";
```
Comment (EvolveArt) [inline crates/pallets/starknet/src/tests/constants.rs:35]: add comment

File: crates/pallets/starknet/src/tests/mod.rs [hunk 2]
```diff
@@ -223,3 +228,56 @@ fn get_invoke_openzeppelin_dummy() -> Transaction {
     }
     .from_invoke(Starknet::chain_id())
 }
+
+pub fn get_declare_dummy(account_type: AccountType) -> DeclareTransaction {
```
Comment (EvolveArt) [inline crates/pallets/starknet/src/tests/mod.rs:232]: rust doc

File: crates/pallets/starknet/src/tests/mod.rs [hunk 3]
```diff
@@ -223,3 +228,56 @@ fn get_invoke_openzeppelin_dummy() -> Transaction {
...
+        class_hash: erc20_class_hash,
+        compiled_class_hash: None,
+        nonce: Felt252Wrapper::ZERO,
+        max_fee: Felt252Wrapper::from(u128::MAX),
+        signature: bounded_vec!(),
+        is_query: false,
+    }
+}
+
+pub fn get_deploy_account_dummy(salt: &str, account_type: AccountType) -> DeployAccountTransaction {
```
Comment (EvolveArt) [inline crates/pallets/starknet/src/tests/mod.rs:252]: rustdoc

File: crates/pallets/starknet/src/tests/mod.rs [hunk 4]
```diff
@@ -223,3 +228,56 @@ fn get_invoke_openzeppelin_dummy() -> Transaction {
...
+        )
+        .unwrap(),
+        nonce: Felt252Wrapper::ZERO,
+        max_fee: Felt252Wrapper::from(u128::MAX),
+        signature: bounded_vec!(),
+        is_query: false,
+    }
+}
+
+pub fn set_infinite_tokens<T: Config>(address: Felt252Wrapper) {
```
Comment (EvolveArt) [inline crates/pallets/starknet/src/tests/mod.rs:274]: rustdoc

File: crates/pallets/starknet/src/tests/mod.rs (hunk 3 above)
Comment (EvolveArt) [reply crates/pallets/starknet/src/tests/mod.rs:252]: why not salt as `Felt252Wrapper` ?

**Review by EvolveArt — CHANGES_REQUESTED** [review]
left few comments, ideally lets merge #868 first cause it comes up with a nice refractor of `MockRuntime` configs

File: crates/runtime/src/pallets.rs [hunk 5]
```diff
@@ -44,6 +44,7 @@ impl pallet_starknet::Config for Runtime {
     type DisableTransactionFee = ConstBool<false>;
     #[cfg(feature = "disable-transaction-fee")]
     type DisableTransactionFee = ConstBool<true>;
+    type EnableNonceValidation = ConstBool<true>;
```
Comment (EvolveArt) [inline crates/runtime/src/pallets.rs:47]: miss the feature flags no?

**Review by EvolveArt — CHANGES_REQUESTED** [review]
lgtm

File: crates/pallets/starknet/src/lib.rs [hunk 6]
```diff
@@ -158,6 +158,9 @@ pub mod pallet {
         /// A bool to disable transaction fees and make all transactions free
         #[pallet::constant]
         type DisableTransactionFee: Get<bool>;
+        /// A bool to enable/disable Nonce validation
+        /// It has been requested by certain apps to allow for nonce validation to be disabled
```
Comment (LucasLvy) [inline crates/pallets/starknet/src/lib.rs:162]: Remove this comment

File: crates/primitives/starknet/src/transaction/mod.rs [hunk 7]
```diff
@@ -560,7 +561,7 @@ impl Transaction {
                 );
 
                 // Update nonce
-                Self::handle_nonce(state, &account_context, self.is_query)?;
+                Self::handle_nonce(state, disable_nonce_validation, &account_context, self.is_query)?;
```
Comment (LucasLvy) [inline crates/primitives/starknet/src/transaction/mod.rs:564]: Don't call that function if we don't validate the nonce

File: crates/primitives/starknet/src/transaction/mod.rs [hunk 8]
```diff
@@ -614,7 +615,7 @@ impl Transaction {
...
```
Comment (LucasLvy) [inline crates/primitives/starknet/src/transaction/mod.rs:618]: same here

File: crates/primitives/starknet/src/transaction/mod.rs [hunk 9]
```diff
@@ -646,7 +647,7 @@ impl Transaction {
...
```
Comment (LucasLvy) [inline crates/primitives/starknet/src/transaction/mod.rs:650]: same

File: crates/primitives/starknet/src/transaction/mod.rs [hunk 10]
```diff
@@ -691,13 +692,15 @@ impl Transaction {
...
     /// * `state` - The state to handle the nonce on
+    /// * `nonce_validation` - Whether to validate the nonce
     /// * `account_tx_context` - The transaction context for the account
     ///
     /// # Returns
     ///
     /// * `TransactionExecutionResult<()>` - The result of the nonce handling
     pub fn handle_nonce(
         state: &mut dyn State,
+        disable_nonce_validation: bool,
```
Comment (LucasLvy) [inline crates/primitives/starknet/src/transaction/mod.rs:703]: no need for that then

**Review by LucasLvy — CHANGES_REQUESTED** [review]

File: crates/pallets/starknet/src/lib.rs [hunk 11]
```diff
@@ -726,6 +740,7 @@ pub mod pallet {
                 &mut BlockifierStateAdapter::<T>::default(),
                 &block_context,
                 TxType::L1Handler,
+                false,
```
Comment (LucasLvy) [inline crates/pallets/starknet/src/lib.rs:743]: think you can set it to true as we don't use the nonce for those transactions

File: crates/pallets/starknet/src/lib.rs [hunk 12]
```diff
@@ -658,11 +669,14 @@ pub mod pallet {
             // Get current block context
             let block_context = Self::get_block_context();
 
+            let disable_nonce_validation = T::DisableNonceValidation::get();
```
Comment (LucasLvy) [inline crates/pallets/starknet/src/lib.rs:672]: don't use an intermediate variable

**Review by LucasLvy — CHANGES_REQUESTED** [review]

**Review by abdelhamidbakhta — APPROVED** [review]
lgtm

**Pull request merged**